File: frontend/src/app/services/house.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { House, HouseRequest, HouseFilter } from '../models/house';

const API_URL = 'http://localhost:8080/api/houses';

@Injectable({
  providedIn: 'root'
})
export class HouseService {

  constructor(private http: HttpClient) { }

  searchHouses(filter: HouseFilter): Observable<House[]> {
    let params = new HttpParams();
    if (filter.keyword) params = params.set('keyword', filter.keyword);
    if (filter.city) params = params.set('city', filter.city);
    if (filter.minPrice) params = params.set('minPrice', filter.minPrice);
    if (filter.maxPrice) params = params.set('maxPrice', filter.maxPrice);
    if (filter.minArea) params = params.set('minArea', filter.minArea);
    if (filter.maxArea) params = params.set('maxArea', filter.maxArea);
    if (filter.bedroom) params = params.set('bedroom', filter.bedroom);
    if (filter.houseType) params = params.set('houseType', filter.houseType);
    if (filter.decoration) params = params.set('decoration', filter.decoration);

    return this.http.get<House[]>(`${API_URL}/search`, { params });
  }

  getHouseById(id: number): Observable<House> {
    return this.http.get<House>(`${API_URL}/${id}`);
  }

  getMyHouses(): Observable<House[]> {
    return this.http.get<House[]>(`${API_URL}/my`);
  }

  createHouse(house: HouseRequest): Observable<House> {
    return this.http.post<House>(API_URL, house);
  }

  updateHouse(id: number, house: HouseRequest): Observable<House> {
    return this.http.put<House>(`${API_URL}/${id}`, house);
  }

  deleteHouse(id: number): Observable<any> {
    return this.http.delete(`${API_URL}/${id}`);
  }

  toggleStatus(id: number): Observable<any> {
    return this.http.post(`${API_URL}/${id}/toggle-status`, {});
  }
}

File: backend/src/main/java/com/houserent/service/HouseService.java
package com.houserent.service;

import com.houserent.dto.HouseRequest;
import com.houserent.entity.House;
import com.houserent.entity.User;
import com.houserent.repository.HouseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.criteria.Predicate;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
public class HouseService {

    @Autowired
    private HouseRepository houseRepository;

    @Autowired
    private AuthService authService;

    @Transactional
    public House createHouse(HouseRequest request) {
        User currentUser = authService.getCurrentUser();
        if (currentUser == null) {
            throw new RuntimeException("用户未登录");
        }
        if (!"LANDLORD".equals(currentUser.getRole())) {
            throw new RuntimeException("只有房东可以发布房源");
        }

        House house = new House();
        house.setLandlord(currentUser);
        mapRequestToHouse(request, house);
        house.setStatus("PUBLISHED");

        return houseRepository.save(house);
    }

    @Transactional
    public House updateHouse(Long id, HouseRequest request) {
        House house = houseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("房源不存在"));

        User currentUser = authService.getCurrentUser();
        if (currentUser == null || !house.getLandlord().getId().equals(currentUser.getId())) {
            throw new RuntimeException("无权限修改此房源");
        }

        mapRequestToHouse(request, house);
        return houseRepository.save(house);
    }

    @Transactional
    public void toggleHouseStatus(Long id) {
        House house = houseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("房源不存在"));

        User currentUser = authService.getCurrentUser();
        if (currentUser == null || !house.getLandlord().getId().equals(currentUser.getId())) {
            throw new RuntimeException("无权限操作此房源");
        }

        if ("PUBLISHED".equals(house.getStatus())) {
            house.setStatus("OFFLINE");
        } else {
            house.setStatus("PUBLISHED");
        }

        houseRepository.save(house);
    }

    @Transactional
    public void deleteHouse(Long id) {
        House house = houseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("房源不存在"));

        User currentUser = authService.getCurrentUser();
        if (currentUser == null || !house.getLandlord().getId().equals(currentUser.getId())) {
            throw new RuntimeException("无权限删除此房源");
        }

        houseRepository.delete(house);
    }

    public Optional<House> getHouseById(Long id) {
        return houseRepository.findById(id);
    }

    public List<House> getMyHouses() {
        User currentUser = authService.getCurrentUser();
        if (currentUser == null) {
            throw new RuntimeException("用户未登录");
        }
        return houseRepository.findByLandlordOrderByCreatedAtDesc(currentUser);
    }

    public List<House> searchHouses(String keyword, String city, Integer minPrice, Integer maxPrice,
                                      Integer minArea, Integer maxArea, Integer bedroom,
                                      String houseType, String decoration) {
        Specification<House> spec = (root, query, criteriaBuilder) -> {
            List<Predicate> predicates = new ArrayList<>();

            predicates.add(criteriaBuilder.equal(root.get("status"), "PUBLISHED"));

            if (keyword != null && !keyword.isEmpty()) {
                predicates.add(criteriaBuilder.or(
                        criteriaBuilder.like(root.get("title"), "%" + keyword + "%"),
                        criteriaBuilder.like(root.get("district"), "%" + keyword + "%"),
                        criteriaBuilder.like(root.get("address"), "%" + keyword + "%")
                ));
            }

            if (city != null && !city.isEmpty()) {
                predicates.add(criteriaBuilder.equal(root.get("city"), city));
            }

            if (minPrice != null) {
                predicates.add(criteriaBuilder.greaterThanOrEqualTo(root.get("price"), BigDecimal.valueOf(minPrice)));
            }

            if (maxPrice != null) {
                predicates.add(criteriaBuilder.lessThanOrEqualTo(root.get("price"), BigDecimal.valueOf(maxPrice)));
            }

            if (minArea != null) {
                predicates.add(criteriaBuilder.greaterThanOrEqualTo(root.get("area"), minArea));
            }

            if (maxArea != null) {
                predicates.add(criteriaBuilder.lessThanOrEqualTo(root.get("area"), maxArea));
            }

            if (bedroom != null) {
                predicates.add(criteriaBuilder.equal(root.get("bedroom"), bedroom));
            }

            if (houseType != null && !houseType.isEmpty()) {
                predicates.add(criteriaBuilder.equal(root.get("houseType"), houseType));
            }

            if (decoration != null && !decoration.isEmpty()) {
                predicates.add(criteriaBuilder.equal(root.get("decoration"), decoration));
            }

            query.orderBy(criteriaBuilder.desc(root.get("createdAt")));
            return criteriaBuilder.and(predicates.toArray(new Predicate[0]));
        };

        return houseRepository.findAll(spec);
    }

    private void mapRequestToHouse(HouseRequest request, House house) {
        house.setTitle(request.getTitle());
        house.setDescription(request.getDescription());
        house.setProvince(request.getProvince());
        house.setCity(request.getCity());
        house.setDistrict(request.getDistrict());
        house.setAddress(request.getAddress());
        house.setPrice(request.getPrice());
        house.setArea(request.getArea());
        house.setBedroom(request.getBedroom());
        house.setLivingRoom(request.getLivingRoom());
        house.setBathroom(request.getBathroom());
        house.setOrientation(request.getOrientation());
        house.setFloor(request.getFloor());
        house.setDecoration(request.getDecoration());
        house.setHouseType(request.getHouseType());
        house.setImages(request.getImages());
        house.setFacilities(request.getFacilities());
    }
}
